src/simulators/qasm/qasm_controller.hpp

```diff
@@ -73,6 +79,14 @@ class QasmController : public Base::Controller {
   // Clear the current config
   void virtual clear_config() override;
 
+  // Add circuit optimization
+  template <typename Type>
+  inline auto add_optimization(Type&& opt)
```

Comment:
This should be added to the `Base::Controller` class so it can be used by all the controller subclasses. I would also rename it to be `add_circuit_optimization` to be specific about what optimizations we are talking about